otherlibs/stdune/src/appendable_list.ml

```diff
@@ -1,13 +1,57 @@
-type 'a t = 'a list -> 'a list
+type 'a t =
+  | Empty
+  | Singleton of 'a
+  | Cons of 'a * 'a t
+  | Append of 'a t * 'a t
+  | List of 'a list
+  | Concat of 'a t list
```

Comment:
How about
```suggestion
  | List of 'a * 'a * 'a list
  | Concat of 'a t * 'a t * 'a t list
```
to rule out lists shorter than 2 elements? We've got no use for them, I think.

I wanted to write: 

> This has the advantage that it's obvious from the type that there is only one way to create an empty list -- using the `Empty` constructor.

but then realised that it's unfortuantely not true due to the `Concat Empty Empty []` case, which is a bit sad.

Reply:
We've got no use for short lists, but these short lists aren't going to be more efficient. There's no real difference between `List (x :: y :: xs)` and `Cons (x, Cons (y, List xs))`.

> but then realised that it's unfortuantely not true due to the Concat Empty Empty Empty [] case, which is a bit sad.

Fixed this. Now we always use `Empty` for the empty list.

Reply:
I don't quite follow your example with lists. Surely `List (x, y, xs)` is a better representation than either the one you get from standard lists or from `Cons` cells. It we've already gone through the trouble of eliminating short list cases, it seems nice to benefit from this in the types and in terms of removing a couple of unnecessary memory indirections.

Reply:
Ah yeah, indeed. I misread your suggestion. Modified `List` to be a little more efficient. We could also introduce a constructor for `Tuple of 'a * 'a` (in the same spirit as `Singleton`). To really make two elements as efficient as possible.

Reply:
Thanks! Not sure how useful `Tuple` is going to be in practice but yeah, we could add it too.

Note that `Concat` is similar to `List`: we can guarantee that it has at least two elements.